examples/02_miscellaneous/plot_1_memory_usage_callback.py

```diff
@@ -0,0 +1,170 @@
+# %%
+"""
+Memory usage at each BCD iteration
+==================================
+
+In this example, we use a callback function to monitor memory usage
+at each iteration of the block-coordinate descent (BCD) algorithm.
+This can be useful to detect memory leaks, or to check that the
+memory usage is not too high for a given device.
+"""
+
+import re
+
+from functools import partial
+
+import matplotlib.pyplot as plt
+import torch
+
+from fugw.mappings import FUGW
+from fugw.utils import _init_mock_distribution
+from rich.console import Console
+from rich.table import Table
+
+# %%
+torch.manual_seed(0)
+
+n_points_source = 50
+n_points_target = 40
+n_features_train = 2
+n_features_test = 2
+
+# %%
+# Let us generate random training data for the source and target distributions
+_, source_features_train, source_geometry, source_embeddings = (
+    _init_mock_distribution(n_features_train, n_points_source)
+)
+_, target_features_train, target_geometry, target_embeddings = (
+    _init_mock_distribution(n_features_train, n_points_target)
+)
+
+# %%
+# Features and geometries should be normalized before calling the solver
+source_features_train_normalized = source_features_train / torch.linalg.norm(
+    source_features_train, dim=1
+).reshape(-1, 1)
+target_features_train_normalized = target_features_train / torch.linalg.norm(
+    target_features_train, dim=1
+).reshape(-1, 1)
+
+source_geometry_normalized = source_geometry / source_geometry.max()
+target_geometry_normalized = target_geometry / target_geometry.max()
+
+
+# %%
+# We define a function to check memory usage at each iteration of the BCD
+# algorithm, as well as util functions to print relevant information.
+def is_sparse(t):
+    return str(t.layout).find("sparse") >= 0
+
+
+def str_size(s):
+    m = re.match(r"torch\.Size\(\[(.*)\]\)", str(s))
+    return f"{m.group(1)}"
+
+
+def str_mem(mem, unit="KB"):
+    if unit == "KB":
+        return f"{mem / 1024:,.3f} KB"
+    elif unit == "MB":
+        return f"{mem / 1024 ** 2:,.3f} MB"
+
+
+def check_memory_usage(locals, device=torch.device("cpu")):
```

Comment:
I feel that it could be useful to create a ``mem_utils.py`` file to store those different technical functions. It would allow us to use them for quick monitoring purposes and they can still be imported in this file. Memory management can be hard to understand therefore I would advise not to include these functions directly in the example.

If you still want to show the power user how to use implement a custom memory callback function, I would advise to create another example named ``advanced_memory_monitoring.py`` to inform the fugw newcomer browsing examples that this file is more technical.

Reply:
This is a fair point, but I am a bit reluctant to add too many things to this package. In particular, if we add public util functions, we have to test them, document them, maintain them, etc.
If you think you'd reuse them elsewhere, we could add them as private functions to the `utils` module, but they would still need testing. I think we should merge this first and maybe integrate these functions later on if time allows.

Moreover, I agree with you that we should have another less challenging example for callbacks ; I was thinking of adding one that computes correlation between validation features at each BCD step. Do you think you would want to handle this ? :blush:

Reply:
I just modified the new example to better describe what is going on.

Reply:
Sure ! I already have it lying around so it should be a simple addition.